tutor/commands/upgrade/common.py

```diff
@@ -66,3 +66,55 @@ def get_mongo_upgrade_parameters(
     mv '/openedx/data/ora2/SET-ME-PLEASE (ex. bucket-name)' /openedx/data/ora2/openedxuploads
 fi
 """
+
+def get_mysql_change_charset_query(
+    database: str,
+    charset: str,
+    collation: str,
+    query_to_append: str,
+) -> None:
+    """
+    Helper function to generate the mysql query to upgrade the charset and collation of tables
+    
+    Utilized in the `tutor local do change-charset-collation` command
+    """
+    return f"""
+            DROP PROCEDURE IF EXISTS UpdateTables;
+            DELIMITER $$
+
+            CREATE PROCEDURE UpdateTables()
+            BEGIN
+
+                DECLARE done INT DEFAULT FALSE;
+                DECLARE _table_name VARCHAR(255);
+                DECLARE cur CURSOR FOR
+                        SELECT table_name FROM information_schema.tables
+                        WHERE table_schema = '{database}' AND table_type = 'BASE TABLE' {query_to_append};
+                DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;
+                Select "The following tables will have their charset and collation updated:";
+                OPEN cur;
+                tables_loop: LOOP
+                    FETCH cur INTO _table_name;
+
+                    IF done THEN
+                    LEAVE tables_loop;
+                    END IF;
+
+                    SET FOREIGN_KEY_CHECKS = 0;
+                    Select _table_name;
+                    SET @statement = CONCAT('ALTER TABLE ', _table_name, ' CONVERT TO CHARACTER SET {charset} COLLATE {collation};'); 
```

Comment:
Will this change columns that currently use `utf8mb4_bin` collation to be `utf8mb4_unicode_ci`? If so, that's a problem. The relatively small number of columns that use `utf8mb4_bin` are usually doing so because they explicitly want the values to be case sensitive, because case insensitivity on things like course keys has given us a lot of grief in the past.

Reply:
To clarify that last statement, an example was that MongoDB and the Python layer of CourseKeys are case sensitive, but the key stored in our CourseOverview model is not case sensitive. So at one point we had two course runs in production that had keys that differed only by case (because someone had misspelled when doing the initial creation), and updates to one could clobber the values for the other because they were fighting over the same row in CourseOverview for updates. (The inactive course run was still getting automated push updates from an external system.)

I realize that fixing CourseOverview is well outside the scope of this work, and I'm not trying to suggest we tackle that here. I only put it out there as an example of why apps might want to explicitly specify a sensitive field for storing things like identifier keys.

Reply:
> Will this change columns that currently use utf8mb4_bin collation to be utf8mb4_unicode_ci?

It seems this is the case. I think we can add a condition that columns with the `utf8mb4_bin` collation should not be upgraded. Any other specific charsets or collations that we should look out for here? 

Or we can just upgrade all the columns that are using `utf8mb3_general_ci` so that if some other column has their charset/collation fixed by the model in the future, we don't have to update the check for them.

Reply:
Based on [your findings](https://github.com/overhangio/tutor/issues/938#issuecomment-2090059383), it seems like the only forced collations that are not `utf8mb4_unicode_ci` are:

* `utf8mb3_bin`: This should be mapped into `utf8mb4_bin`, since that's the logically equivalent case-sensitive collation.
* `utf8mb4_bin`: This should be preserved as-is, as discussed.
* `utf8mb4_general_ci`: This would be mapped to `utf8mb4_unicode_ci`, but it looks like it only exists for the `bundles` app, and that app is unused in Redwood and removed entirely just after the Redwood cut, so it's fine to ignore or update it, depending on what's more convenient in the code.

Reply:
I've updated the command to perform the following mappings instead:

- `utf8mb3_general_ci` to `utf8mb4_unicode_ci`
- `utf8mb3_*` to `utf8mb4_*` meaning that the `utf8mb3_bin` collations are converted to `utf8mb4_bin` now.

> utf8mb4_bin: This should be preserved as-is, as discussed.

We ignore any columns that already have the `utf8mb4` charset.

> utf8mb4_general_ci: This would be mapped to utf8mb4_unicode_ci, but it looks like it only exists for the bundles app, and that app is unused in Redwood and removed entirely just after the Redwood cut, so it's fine to ignore or update it, depending on what's more convenient in the code.

Yup, trying it out on the nightly release, the bundles app was not there. Therefore, I've ignored it.